base/reflection.jl

```diff
@@ -1421,22 +1426,21 @@ function infer_effects(@nospecialize(f), @nospecialize(types=default_tt(f));
         argtypes = Any[types.parameters...]
         rt = Core.Compiler.builtin_tfunction(interp, f, argtypes, nothing)
         return Core.Compiler.builtin_effects(f, argtypes, rt)
-    else
-        effects = Core.Compiler.EFFECTS_TOTAL
-        matches = _methods(f, types, -1, world)::Vector
-        if isempty(matches)
-            # this call is known to throw MethodError
-            return Core.Compiler.Effects(effects; nothrow=false)
-        end
-        for match in matches
-            match = match::Core.MethodMatch
-            frame = Core.Compiler.typeinf_frame(interp,
-                match.method, match.spec_types, match.sparams, #=run_optimizer=#false)
-            frame === nothing && return Core.Compiler.Effects()
-            effects = Core.Compiler.merge_effects(effects, frame.ipo_effects)
-        end
-        return effects
     end
+    effects = Core.Compiler.EFFECTS_TOTAL
+    matches = _methods(f, types, -1, world)::Vector
+    if isempty(matches)
```

Comment:
shouldn't this need to check for the `ambiguous` flag and `fully_covers` flag (like inlining) to be entirely correct?

Reply:
Yes, we should account for that. This `isempty`-based check is not sound.